### PR TITLE
[PR3] Add manual NAT settings

### PR DESCRIPTION
**Pull request by matkt**
Signed-off-by: Karim TAAM karim.t2am@gmail.com

## PR description

Ability to explicitly configure the external IP to broadcast without regards to NAT or other considerations.

## Changes list

- Adding a new NatMethod `Manual`
- Modification of the `NatService` builder. It will take directly the `NatManager` and not the method
- Creation of a `ManualNatManager` class that inherits from the` AbstractNatManager` class.

This class will define the ports and IP to broadcast. Here is the list of command line parameters that will be used
- For ports it's `p2p-port` and` rpc-http-port` parameters
- For the IP it is `p2p-host`

### CODE REVIEW

**Comment by matkt** [comment]
Hi @RatanRSur the PR3 is ready to be reviewed

File: nat/src/main/java/org/hyperledger/besu/nat/NatService.java [hunk 1]
```diff
@@ -170,27 +169,16 @@ public static Builder builder() {
...
   public static class Builder {
-    private NatMethod natMethod = NatMethod.NONE;
+    private Optional<NatManager> natManager = Optional.empty();
 
-    public Builder natMethod(final NatMethod natMethod) {
-      checkNotNull(natMethod);
-      this.natMethod = natMethod;
+    public Builder natManager(final Optional<NatManager> natManager) {
+      checkNotNull(natManager);
+      this.natManager = natManager;
```
Comment (RatanRSur) [inline nat/src/main/java/org/hyperledger/besu/nat/NatService.java]: You can inline `checkNotNull`

File: nat/src/main/java/org/hyperledger/besu/nat/manual/ManualNatManager.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  private final int p2pPort;
+  private final int rpcHttpPort;
+  private final List<NatPortMapping> forwardedPorts;
+
+  public ManualNatManager(final String advertisedHost, final int p2pPort, final int rpcHttpPort) {
+    super(NatMethod.MANUAL);
+    this.advertisedHost = advertisedHost;
+    this.p2pPort = p2pPort;
+    this.rpcHttpPort = rpcHttpPort;
+    forwardedPorts = buildForwardedPorts();
```
Comment (RatanRSur) [inline nat/src/main/java/org/hyperledger/besu/nat/manual/ManualNatManager.java]: `this.` just for consistency :)

File: nat/src/main/java/org/hyperledger/besu/nat/manual/ManualNatManager.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+              p2pPort),
+          new NatPortMapping(
+              NatServiceType.JSON_RPC,
+              NetworkProtocol.TCP,
+              internalHost,
+              advertisedHost,
+              rpcHttpPort,
+              rpcHttpPort));
+    } catch (Exception e) {
+      LOG.info("Failed to create forwarded port list");
```
Comment (RatanRSur) [inline nat/src/main/java/org/hyperledger/besu/nat/manual/ManualNatManager.java]: What do you think about adding the specific exception to this log?

File: nat/src/main/java/org/hyperledger/besu/nat/NatService.java [hunk 4]
```diff
@@ -170,27 +169,16 @@ public static Builder builder() {
...
-    /** Initialize the current NatManager. */
-    private Optional<NatManager> buildNatManager(final NatMethod givenNatMethod) {
-      switch (givenNatMethod) {
-        case UPNP:
-          return Optional.of(new UpnpNatManager());
-        case NONE:
-        default:
-          return Optional.empty();
-      }
+      return new NatService(natManager);
```
Comment (RatanRSur) [inline nat/src/main/java/org/hyperledger/besu/nat/NatService.java]: Is a builder required if there's only one parameter?

**Comment by RatanRSur** [comment]
Regarding `start` and `stop` in `NatManagerService`, can we get into a case where we try to start but there is no nat environment? If we can get into this state, should we log that in `start` and `stop`?

**Comment by matkt** [comment]
> Regarding `start` and `stop` in `NatManagerService`, can we get into a case where we try to start but there is no nat environment? If we can get into this state, should we log that in `start` and `stop`?

`start` and` stop` methods are called each time. The only difference is that if we are not in a NAT environment nothing will happen because the `NatService` will check before calling the manager. Indeed we can add a log to specify that no service is started. I propose this modification

- Current implementation

```java
public void start() {
	if (isNatEnvironment()) {
	  try {
		getNatManager().orElseThrow().start();
	  } catch (Exception e) {
		LOG.warn("Caught exception while trying to start the manager or service", e);
	  }
	}
}

public void stop() {
	if (isNatEnvironment()) {
	  try {
		getNatManager().orElseThrow().stop();
	  } catch (Exception e) {
		LOG.warn("Caught exception while trying to stop the manager or service", e);
	  }
	}
}
```

- New implementation

```java
public void start() {
	if (isNatEnvironment()) {
	  try {
		getNatManager().orElseThrow().start();
	  } catch (Exception e) {
		LOG.warn("Caught exception while trying to start the manager or service", e);
	  }
	} else {
	  LOG.info("No NAT environment detected so no service could be started");
	}
}

public void stop() {
	if (isNatEnvironment()) {
	  try {
		getNatManager().orElseThrow().stop();
	  } catch (Exception e) {
		LOG.warn("Caught exception while trying to stop the manager or service", e);
	  }
	} else {
	  LOG.info("No NAT environment detected so no service could be stopped");
	}
}
```

File: nat/src/main/java/org/hyperledger/besu/nat/NatService.java (hunk 4 above)
Comment (matkt) [reply nat/src/main/java/org/hyperledger/besu/nat/NatService.java]: yes you're right I'm going to delete the builder

File: nat/src/main/java/org/hyperledger/besu/nat/manual/ManualNatManager.java (hunk 3 above)
Comment (matkt) [reply nat/src/main/java/org/hyperledger/besu/nat/manual/ManualNatManager.java]: Done

File: nat/src/main/java/org/hyperledger/besu/nat/manual/ManualNatManager.java (hunk 2 above)
Comment (matkt) [reply nat/src/main/java/org/hyperledger/besu/nat/manual/ManualNatManager.java]: Done

File: nat/src/main/java/org/hyperledger/besu/nat/NatService.java (hunk 1 above)
Comment (matkt) [reply nat/src/main/java/org/hyperledger/besu/nat/NatService.java]: the builder no longer exists

**Comment by RatanRSur** [comment]
After the log modification, looks good to me!

**Review by RatanRSur — APPROVED** [review]

**Pull request merged**